crates/bevy_window/src/window.rs

```diff
@@ -86,36 +136,109 @@ pub enum CompositeAlphaMode {
     Inherit = 4,
 }
 
-impl WindowId {
-    /// Creates a new [`WindowId`].
-    pub fn new() -> Self {
-        WindowId(Uuid::new_v4())
-    }
-    /// The [`WindowId`] for the primary window.
-    pub const fn primary() -> Self {
-        WindowId(Uuid::from_u128(0))
-    }
-    /// Get whether or not this [`WindowId`] is for the primary window.
-    pub fn is_primary(&self) -> bool {
-        *self == WindowId::primary()
-    }
+/// Defines the way a window is displayed
+#[derive(Default, Debug, Component, Clone, Copy, PartialEq, Eq, Reflect)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+#[reflect(Component)]
+pub enum WindowMode {
+    /// Creates a window that uses the given size.
+    #[default]
+    Windowed,
+    /// Creates a borderless window that uses the full size of the screen.
+    BorderlessFullscreen,
+    /// Creates a fullscreen window that will render at desktop resolution. The app will use the closest supported size
+    /// from the given size and scale it to fit the screen.
+    SizedFullscreen,
+    /// Creates a fullscreen window that uses the maximum supported size.
+    Fullscreen,
 }
 
-use crate::CursorIcon;
-use std::fmt;
-
-use crate::raw_handle::RawHandleWrapper;
+/// Define how a window will be created and how it will behave.
+#[derive(Default, Resource, Bundle, Debug, Clone)]
+pub struct WindowBundle {
```

Comment:
I'm not fully sold on the "granular componentization" of Window. From my perspective, it degrades the user experience in a number of areas:

It makes the initial setup of windows weirder and more confusing because things that used to be fields now require wrapper types:

```rust
// Before
let descriptor = WindowDescriptor {
    width: 500.,
    height: 300.,
    title: "I am a window!".to_string(),
    ..default()
}

// After
let bundle = WindowBundle {
    // This isn't a Vec2, its a WindowResolution
    resolution: (500., 300.).into(),
    // this isn't a String, its a WindowTitle
    title: "I am a window!".into(),
    ..default()
}
```

Accessing Window information is now less clear because we need to "summon" each component and autocomplete no longer reveals all available window fields and functions:

```rust
// Before
fn system(windows: Res<Windows>) {
    let primary = windows.primary()
    println!("width: {}, height: {}, title: {}", primary.width(), primary.height(), primary.title());
}

// After
fn system(windows: Query<(&WindowResolution, &WindowTitle)>) {
    let (resolution, title) = windows.single()
    println!("width: {}, height: {}, title: {}", resolution.width(), resolution.height(), primary.title());
}
```

This is especially true for some single-component queries, which feel way less clear to me now:

```rust
// Before
fn system(windows: Res<Windows>) {
    let primary = windows.primary()
    println!("present mode: {}", primary.present_mode());
}

// After
fn system(present_modes: Query<&PresentMode>) {
    let present_mode = present_modes.single()
    println!("present mode: {}", present_mode);
}
```

Note that in the system above, the term "window" doesn't even show up anymore, despite it being a "window" system.

Theres also the problem that all of these "window behaviors" aren't really "separable". Windows will always have a position and a resolution, by definition. Removing WindowResolution and leaving WindowPosition is not valid, but it is currently possible in this PR. Breaking them up into separate "component concepts" kind of breaks component design rules of thumb because they cannot exist independently as separate behaviors.

Switching to a consolidated `Window` component would allow us to solve these problems:

```rust
// Window setup "fields" can be simpler and clearer because they don't need wrapper types
let window = Window {
    requested_width: 500.,
    requested_height: 300.,
    title: "I am a window!".to_string(),
    ..default()
}

/// Accessing multiple window properties doesn't explode the query complexity, and window properties are easily discoverable
fn system(windows: Query<&Window>) {
    let window = windows.single()
    println!("width: {}, height: {}, title: {}", window.width(), window.height(), window.title());
}

// Single-property access retains the "window" concept / terminology
fn system(windows: Query<&Window>) {
    let window = windows.single()
    println!("present mode: {}", window.present_mode());
}

// This removes the window "component behavior" as a whole. There is no way
// To invalidate the window behavior by removing something like WindowPosition
commands.entity(entity).remove::<Window>();
```

But a centralized window creates problems of the "change detection" variety, which seems like the primary motivator for hyper-granular componentization (both here and elsewhere).
It is hard to deny the value of being able to query for `Changed<WindowResolution>`. But for users, checking for window property changes will be less common than
getting and setting them, and we already have window events for change detection use cases.

I can think of two paths for "centralized windows":

1. Use coarse change detection in combination with comparison. The winit window impl would first query for `Changed<Window>` and then compare the current state of the winit window to the changed window (across all fields/properties) to see if a change has occurred. In practice, I think this would be totally fine. Most frames won't have window changes, the number of total windows will always be "small" (as they are "human managed"), and windows aren't particularly "big" or expensive to compare. This cost seems even more ok given that "granular window components" requires scheduling and running ~10 systems every frame, whereas the window-change-comparison would happen in a single system that only evaluates per-field logic when the Window component is changed.
2. Embed "dirty" bits into the Window component and use getters/setters for each field. Window construction would need to either employ the builder pattern or go back to descriptor-based construction (both of which are definitely their own kind of UX degradation). Reflection would probably want to ignore the dirty bits.

There is also the third path of solving the "granular change detection for unified concept" problem in ECS more generally with some new ECS magic. But that is hard, maybe impossible to do well enough, and not something that should block this PR.

Note that I'm still working out how I feel about this / I'm not prescribing anything. I'm just currently slightly leaning toward a unified Window component and I'd like to kick off the conversation.

Reply:
Yeah, I feel the exact same way, making components too granular makes things harder to discover, especially for beginners. I understand why it happens for performance reasons in other context, but here where the vast majority of situations will only have a single window anyway, I really don't think the performance is a concern compared to the worse UX.

Reply:
The UX is one thing, but yeah definitely doesn't feel right that  you can remove components that leave the entity in an unexpected state (like in your removing resolution but not position example).

Reply:
I think I'd be fine with a unified component, I mainly did it for change detection but it should be completely possible to just compare most fields with winit's (or a `PreviousWindow(WindowDescriptor)`), one thing I would warn about this though is that minimization/maximization will probably still need to be split out somehow, since there isn't a really concrete way to tell whether a window is "minimized".

Reply:
Yeah I'm in favor of a unified component here.

Reply:
Cool I think we should roll with "centralized windows option number 1" (or some variant of it). In addition to the reasons stated above, it should also make migrations easier because it will have closer parity with the current WindowDescriptor and Window types. Which makes it a less controversial change in my book. And we can always revisit if the need arises.

Reply:
This should be done now